### PR TITLE
N4509 constexpr atomic<T>::is_always_lock_free

### PR DESCRIPTION
**Pull request by jfbastien**

### CODE REVIEW

File: source/N4509.rst [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+The function ``atomic_is_always_lock_free`` (*29.6*) indicates whether the
+object is always lock-free. In any given program execution, the result of the
+lock-free query shall be consistent for all pointers of the same type.
+
+i.e. the following should never assert:
+
+.. code-block:: c++
+
+  if (atomic<T>::is_lock_free())
+    assert(is_always_lock_free<T>::value);
```
Comment (ogiroux) [inline source/N4509.rst]: This needs to call the member function, not the trait. We're not selling the trait.

File: source/N4509.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+The function ``atomic_is_always_lock_free`` (*29.6*) indicates whether the
+object is always lock-free. In any given program execution, the result of the
+lock-free query shall be consistent for all pointers of the same type.
+
+i.e. the following should never assert:
+
+.. code-block:: c++
+
+  if (atomic<T>::is_lock_free())
```
Comment (ogiroux) [inline source/N4509.rst]: You need an instance to call this because it's a non-static member function.

File: source/N4509.rst [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+:Author: JF Bastien
+:Contact: jfb@google.com
+:Date: 2015-05-05
+:URL: https://github.com/jfbastien/papers/blob/master/source/4509.rst
+:Source: https://github.com/jfbastien/papers/blob/master/source/4509.cc
+
+The current design for ``std::atomic<T>`` affords implementations the critical
+freedom to revert to critical sections when hardware support for atomic
+operations does not meet the size or semantic requirements for the associated
+type ``T``. This preserves C++ support on aging hardware, when developers don't
```
Comment (ogiroux) [inline source/N4509.rst]: The 'when' is awkward.  Another form would be better.  Maybe "supports developers that don't".

File: source/N4509.cc [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+//   assert(is_always_lock_free<T>::value);
+template<typename T> struct is_always_lock_free { enum { value = false }; };
+
+// Implementations must match the C ATOMIC_*_LOCK_FREE macro values.
+template<> struct is_always_lock_free<bool> { enum { value = 2 == ATOMIC_BOOL_LOCK_FREE }; };
+template<> struct is_always_lock_free<char> { enum { value = 2 == ATOMIC_CHAR_LOCK_FREE }; };
+template<> struct is_always_lock_free<char16_t> { enum { value = 2 == ATOMIC_CHAR16_T_LOCK_FREE }; };
+template<> struct is_always_lock_free<char32_t> { enum { value = 2 == ATOMIC_CHAR32_T_LOCK_FREE }; };
+template<> struct is_always_lock_free<wchar_t> { enum { value = 2 == ATOMIC_WCHAR_T_LOCK_FREE }; };
+template<> struct is_always_lock_free<short> { enum { value = 2 == ATOMIC_SHORT_LOCK_FREE }; };
```
Comment (ogiroux) [inline source/N4509.cc]: What about the unsigned versions?

File: source/N4509.cc [hunk 5]
```diff
@@ -0,0 +1,36 @@
+#include <atomic>
+#include <iostream>
+
+// It is implementation-defined what this returns, as long as:
+// if (atomic<T>::is_lock_free())
```
Comment (ogiroux) [inline source/N4509.cc]: This has the same problem as the code sample in the paper.

Comment (ogiroux) [inline source/N4509.cc]: Actually, the trait here needs a bit of expository wrapping to show how it's supposed to be used in std::atomic.  Maybe a prototype for std::atomic::is_always_lock_free in comments.

File: source/N4509.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+improves the situation for HPC while leaving all other uses untouched.
+
+We propose a ``constexpr`` version of ``is_lock_free()`` that is suitable for
+use with SFINAE or ``static_assert``:
+
+.. code-block:: c++
+
+  namespace std {
+    template <class T> struct atomic {
+      constexpr bool is_always_lock_free() const noexcept;
```
Comment (ogiroux) [inline source/N4509.rst]: Make 'em static.

File: source/N4509.rst [hunk 7]
```diff
@@ -0,0 +1,68 @@
+=======================================
+N4509 `constexpr is_always_lock_free()`
```
Comment (ogiroux) [inline source/N4509.rst]: Title needs to change if we make it a const static data member.

**Comment by jfbastien** [comment]
I updated the source code and the paper following @ogiroux's comments, with help from Jeffrey.

**Pull request merged**